File: optionals/med-4.0.0/tests/usecases/python/UsesCase_MEDmesh_10.py
#! /usr/bin/env python
# -*- coding:utf-8 -*-
# /*  This file is part of MED.
#  *
#  *  COPYRIGHT (C) 1999 - 2019  EDF R&D, CEA/DEN
#  *  MED is free software: you can redistribute it and/or modify
#  *  it under the terms of the GNU Lesser General Public License as published by
#  *  the Free Software Foundation, either version 3 of the License, or
#  *  (at your option) any later version.
#  *
#  *  MED is distributed in the hope that it will be useful,
#  *  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  *  GNU Lesser General Public License for more details.
#  *
#  *  You should have received a copy of the GNU Lesser General Public License
#  *  along with MED.  If not, see <http://www.gnu.org/licenses/>.
#  */

# /******************************************************************************
#  *  How to create an unstructured mesh
#  *
#  *  Use case 10 : a 2D unstructured mesh with 15 nodes, 8 triangular cells, 4 triangular cells
#  *  and families
#  *****************************************************************************/

import sys

import med

def doCleanup(fid, status) :
	try :
		med.MEDfileClose(fid)
	except RuntimeError as ex :
		print("ERROR : close file ...\n%s" % ex)
		status = -1
	finally :
		if status != 0 :
			sys.exit(status)

#  max size of meshname : med.MED_NAME_SIZE+1
meshname = "2D unstructured mesh"
spacedim = 2
meshdim = 2
#  max size of axisname and unitname : 2*med.MED_SNAME_SIZE+1
#           12345678901234561234567890123456
axisname = "x               y               "
unitname = "cm              cm              "
coordinates = med.MEDFLOAT([2.,1.,  7.,1.,  12.,1.,  17.,1.,  22.,1.,
							2.,6.,  7.,6.,  12.,6.,  17.,6.,  22.,6.,
							2.,11., 7.,11., 12.,11., 17.,11., 22.,11. ])
nnodes = 15
triaconnectivity = med.MEDINT([ 1,7,6,   2,7,1,  3,7,2,   8,7,3,
								13,7,8, 12,7,13, 11,7,12, 6,7,11 ])
ntria3 = 8
quadconnectivity = med.MEDINT([ 3,4,9,8,    4,5,10,9,
								15,14,9,10, 13,8,9,14 ])
nquad4 = 4
#  max size of familyname : med.MED_NAME_SIZE+1
familyname = "BOUNDARY_VERTICES"
#  max size of groupename : med.MED_LNAME_SIZE+1
groupname = med.MEDCHAR("MESH_BOUNDARY_VERTICES")
familynumbers = med.MEDINT([1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1])

#  MED file creation
try :
	fid = med.MEDfileOpen("UsesCase_MEDmesh_10.med", med.MED_ACC_CREAT)
except RuntimeError as ex :
	print("ERROR : file creation ...\n%s" % ex)
	sys.exit(-1)

#  write a comment in the file
try :
	med.MEDfileCommentWr(fid, "A 2D unstructured mesh : 15 nodes, 12 cells")
except RuntimeError as ex :
	print("ERROR : write file description ...\n%s" % ex)
	doCleanup(fid, -1)

#  mesh creation : a 2D unstructured mesh
try :
	med.MEDmeshCr(fid, meshname, spacedim, meshdim, med.MED_UNSTRUCTURED_MESH,
				"A 2D unstructured mesh", "", med.MED_SORT_DTIT,
				med.MED_CARTESIAN, axisname, unitname)
except RuntimeError as ex :
	print("ERROR : mesh creation ...\n%s" % ex)
	doCleanup(fid, -1)

#  nodes coordinates in a cartesian axis in full interlace mode
#  (X1,Y1, X2,Y2, X3,Y3, ...) with no iteration and computation step
try :
	med.MEDmeshNodeCoordinateWr(fid, meshname, med.MED_NO_DT, med.MED_NO_IT,
							0.0, med.MED_FULL_INTERLACE, nnodes, coordinates)
except RuntimeError as ex :
	print("ERROR : nodes coordinates ...\n%s" % ex)
	doCleanup(fid, -1)

#  cells connectiviy is defined in nodal mode
try :
	med.MEDmeshElementConnectivityWr(fid, meshname, med.MED_NO_DT,med.MED_NO_IT,
							0.0, med.MED_CELL, med.MED_TRIA3, med.MED_NODAL,
							med.MED_FULL_INTERLACE, ntria3, triaconnectivity)
except RuntimeError as ex :
	print("ERROR : triangular cells connectivity ...\n%s" % ex)
	doCleanup(fid, -1)

try :
	med.MEDmeshElementConnectivityWr(fid, meshname, med.MED_NO_DT,med.MED_NO_IT,
							0.0, med.MED_CELL, med.MED_QUAD4, med.MED_NODAL,
							med.MED_FULL_INTERLACE, nquad4, quadconnectivity)
except RuntimeError as ex :
	print("ERROR : quadrangular cells connectivity ...\n%s" % ex)
	doCleanup(fid, -1)

#  create family 0 : by default, all mesh entities family number is 0
#TODO : Etudier la pertinence de définir MED_NO_GROUP comme un MEDCHAR('')
#TODO : A confronter aux types des paramètres axisname, axisunit
try :
	med.MEDfamilyCr(fid, meshname, med.MED_NO_NAME, 0, 0,
				med.MEDCHAR(med.MED_NO_GROUP))
except RuntimeError as ex :
	print("ERROR : family 0 creation ...\n%s" % ex)
	doCleanup(fid, -1)

#  create a family for boundary vertices : by convention a nodes family number is > 0,
#  and an element family number is < 0
try :
	med.MEDfamilyCr(fid, meshname, familyname, 1, 1, groupname)
except RuntimeError as ex :
	print("ERROR : family %s creation ...\n%s" % (familyname, ex))
	doCleanup(fid, -1)

#  write family number for nodes
try :
	med.MEDmeshEntityFamilyNumberWr(fid, meshname, med.MED_NO_DT, med.MED_NO_IT,
						med.MED_NODE, med.MED_NONE, nnodes, familynumbers)
except RuntimeError as ex :
	print("ERROR : nodes family numbers ...\n%s" % ex)
	doCleanup(fid, -1)

#  close MED file
doCleanup(fid, 0)

File: optionals/med-4.0.0/tests/usecases/f/UsesCase_MEDmesh_10.f
C*  This file is part of MED.
C*
C*  COPYRIGHT (C) 1999 - 2019  EDF R&D, CEA/DEN
C*  MED is free software: you can redistribute it and/or modify
C*  it under the terms of the GNU Lesser General Public License as published by
C*  the Free Software Foundation, either version 3 of the License, or
C*  (at your option) any later version.
C*
C*  MED is distributed in the hope that it will be useful,
C*  but WITHOUT ANY WARRANTY; without even the implied warranty of
C*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
C*  GNU Lesser General Public License for more details.
C*
C*  You should have received a copy of the GNU Lesser General Public License
C*  along with MED.  If not, see <http://www.gnu.org/licenses/>.
C*

C******************************************************************************
C * How to create an unstructured mesh
C * Use case 10 : write a 2D unstructured mesh with 15 nodes, 8 triangular
C *               cells, 4 quadrangular cells, and families
C *
C *****************************************************************************
      program UsesCase_MEDmesh_10
C     
      implicit none
      include 'med.hf77'
C
C     
      integer cret
      integer*8 fid

C     space dim, mesh dim      
      integer sdim, mdim
C     axis name, unit name
      character*16 axname(2), unname(2)
C     mesh name, family name, time step unit, file name
      character*64 mname, fyname, dtunit, finame
C     mesh type, sorting type, grid type
      integer mtype, stype, grtype
C     family number, number of group      
      integer fnum, ngro
C     group name
      character*80 gname
C     coordinates, date
      real*8 coords(30), dt
      integer nnodes, ntria3, nquad4
C     triangular and quadrangular cells connectivity
      integer tricon(24), quacon(16)
C     family numbers
      integer fanbrs(15)
C     comment 1, mesh description
      character*200 cmt1, mdesc
C
      parameter (sdim = 2, mdim = 2)
      parameter (mname = "2D unstructured mesh")
      parameter (fyname = "BOUNDARY_VERTICES")
      parameter (dtunit = " ")
      parameter (dt = 0.0d0)
      parameter (finame = "UsesCase_MEDmesh_10.med")
      parameter (gname = "MESH_BOUNDARY_VERTICES")
      parameter (nnodes = 15, ntria3 = 8, nquad4 = 4)
      parameter (cmt1 ="A 2D unstructured mesh : 15 nodes, 12 cells")
      parameter (mtype=MED_UNSTRUCTURED_MESH, stype=MED_SORT_DTIT )
      parameter (mdesc = "A 2D unstructured mesh")
      parameter (grtype=MED_CARTESIAN_GRID)
C
      data axname  /"x" ,"y" /
      data unname  /"cm","cm"/
      data coords /2.,1.,  7.,1.,  12.,1.,  17.,1.,  22.,1.,
     &             2.,6.,  7.,6.,  12.,6.,  17.,6.,  22.,6.,
     &             2.,11., 7.,11., 12.,11., 17.,11., 22.,11./
      data tricon /1,7,6,   2,7,1,  3,7,2,   8,7,3,   
     &             13,7,8, 12,7,13, 11,7,12, 6,7,11/
      data quacon /3,4,9,8,    4,5,10,9, 
     &             15,14,9,10, 13,8,9,14/
      data fanbrs /1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1/
C 
C
C     file creation
      call mfiope(fid,finame,MED_ACC_CREAT,cret)
      if (cret .ne. 0 ) then
         print *,'ERROR : file creation'
         call efexit(-1)
      endif  
C
C
C     write a comment in the file
      call mficow(fid,cmt1,cret)
      if (cret .ne. 0 ) then
         print *,'ERROR : write file description'
         call efexit(-1)
      endif
C
C
C     mesh creation : a 2D unstructured mesh
      call mmhcre(fid, mname, sdim, mdim, mtype, mdesc, dtunit,
     &            stype, grtype, axname, unname, cret)
      if (cret .ne. 0 ) then
         print *,'ERROR : mesh creation'
         call efexit(-1)
      endif
C
C
C     nodes coordinates in a cartesian axis in full interlace mode
C     (X1,Y1, X2,Y2, X3,Y3, ...) with no iteration and computation step
      call mmhcow(fid,mname,MED_NO_DT,MED_NO_IT,dt,
     &            MED_FULL_INTERLACE,nnodes,coords,cret)
      if (cret .ne. 0 ) then
         print *,'ERROR : write nodes coordinates description'
         call efexit(-1)
      endif
C
C
C     cells connectiviy is defined in nodal mode
      call mmhcyw(fid,mname,MED_NO_DT,MED_NO_IT,dt,MED_CELL,
     &            MED_TRIA3,MED_NODAL,MED_FULL_INTERLACE,
     &            ntria3,tricon,cret)
      if (cret .ne. 0 ) then
         print *,'ERROR : triangular cells connectivity'
         call efexit(-1)
      endif 
      call mmhcyw(fid,mname,MED_NO_DT,MED_NO_IT,dt,MED_CELL,
     &            MED_QUAD4,MED_NODAL,MED_FULL_INTERLACE,
     &            nquad4,quacon,cret)
      if (cret .ne. 0 ) then
         print *,'ERROR : quadrangular cells connectivity'
         call efexit(-1)
      endif
C
C
C     create family 0 : by default, all mesh entities family number is 0
      call mfacre(fid,mname,MED_NO_NAME,0,0,MED_NO_GROUP,cret)
      if (cret .ne. 0 ) then
         print *,'ERROR : create family 0'
         call efexit(-1)
      endif
C
C
C     create a family for boundary vertices : by convention a nodes family number is > 0,
C     and an element family number is < 0
      fnum = 1
      ngro = 1
      call mfacre(fid, mname, fyname, fnum, ngro, gname, cret)
      if (cret .ne. 0 ) then
         print *,'ERROR : create family 0'
         call efexit(-1)
      endif
C
C
C     write family number for nodes 
      call mmhfnw(fid, mname, MED_NO_DT, MED_NO_IT, MED_NODE, MED_NONE,
     &            nnodes, fanbrs, cret)
      if (cret .ne. 0 ) then
         print *,'ERROR : nodes family numbers ...'
         call efexit(-1)
      endif
C
C
C     close file
      call mficlo(fid,cret)
      if (cret .ne. 0 ) then
         print *,'ERROR :  close file'
         call efexit(-1)
      endif        
C
C
C
      end
C
